### ISSUE TITLE
Model evaluation not using cross validation

### ISSUE THREAD

**Issue by mrendleman** [larskotthoff/autoweka#50]
When Auto-WEKA reported the performance of its best classifier on my data, it looked great compared to other methods I've tried in the past: 87% accuracy, 0.957 ROC Area. To validate this result, I took the same feature selection and classification methods and tried them outside of Auto-WEKA (CfsSubsetEval with GreedyStepwise search method, meta.Bagging with lazy.IBk), making sure to set all parameters identical to those reported by Auto-WEKA.

When doing 10-fold cross-validation on this Bagging of IBk, the performance was nowhere near that reported by Auto-WEKA (71%, 0.719 ROC Area). However, when I performed the testing on the training data ("Use training set"), WEKA reported the same improved performance (87%, 0.957 ROC area). I also checked the other reported values (confusion matrix, TP/FP rates, precision, recall, F-measure, PRC-area, etc) and they were identical.

This suggests that Auto-WEKA isn't actually performing the 10-fold cross-validation that the documentation suggests it is when evaluating classifiers, but rather testing on the training data instead.

Trying with the Iris dataset supplied with Weka confirms the issue:
Auto-WEKA reports 98% accuracy
The same settings (default simple logistic w/ no feature selection), 10-fold CV reports 94%
The same settings testing on training data reports 98%

I'm not totally certain I haven't misunderstood something, but this seems like a glaring problem if I haven't.

**Comment by larskotthoff** [comment]
The evaluation you're referring to is indeed done on the training set. It is performed by WEKA (not Auto-WEKA!) automatically and there's no way to disable it. I agree that it is very misleading, but there's nothing that can be done in Auto-WEKA about it.

Just before the detailed WEKA evaluation at the end, you can see Auto-WEKA's performance estimate (based on the 10 fold CV). This is more realistic.

**Comment by mrendleman** [comment]
Ah, I see. So the only metric I should be looking at in Auto-WEKA is located higher up in the output, at

```
metric: errorRate
estimated errorRate: <some number>
```

That's fairly unintuitive, especially for users familiar with WEKA and/or new to Auto-WEKA. Maybe something could be added to the documentation about that, but for now at least this issue page might help people out.

**Comment by larskotthoff** [comment]
I agree -- it's really unfortunate. I'll try to make this clearer in the documentation.

**Comment by federicoparra** [comment]
Hi, actually when using auto-weka, even those metrics, 
metric: errorRate
estimated errorRate: <some number>
are overblown. I get often .9 if the metric is correlation, which is nowhere the reality of the model being built.

Example output:

Auto-WEKA result:
best classifier: weka.classifiers.functions.LinearRegression
arguments: [-S, 0, -R, 1.9515327162651044]
attribute search: null
attribute search arguments: []
attribute evaluation: null
attribute evaluation arguments: []
metric: correlationCoefficient
estimated correlationCoefficient: 0.9999520216650307
training time on evaluation dataset: 0.232 seconds

You can use the chosen classifier in your own code as follows:

Classifier classifier = AbstractClassifier.forName("weka.classifiers.functions.LinearRegression", new String[]{"-S", "0", "-R", "1.9515327162651044"});
classifier.buildClassifier(instances);


Correlation coefficient                  0.7247
Mean absolute error                     12.3586
Root mean squared error                 15.3731
Relative absolute error                 67.2423 %
Root relative squared error             68.9104 %
Total Number of Instances              259     


Temporary run directories:
C:\Users\feder\AppData\Local\Temp\autoweka6880147379316683817\


For better performance, try giving Auto-WEKA more time.
Tried 45 configurations; to get good results reliably you may need to allow for trying thousands of configurations.

**Comment by larskotthoff** [comment]
The initially reported correlation coefficient is what Auto-WEKA determines internally and the later one is on the test set you provide. Depending on the split you're using and the composition of the dataset, those may well be quite different.

**Comment by federicoparra** [comment]
Hi @larskotthoff , the first and the last coefficients are both based on the full training sample, i.e., no folds.
Both 0.99xxxx and 0.7 are not the result in any of the folds but the result of using the model on the full dataset.
In other words there is no result based on 10 folds.

**Comment by larskotthoff** [comment]
Ok, if you're running Auto-WEKA as suggested in the manual, the first estimate is based on Auto-WEKA's internal cross validation (i.e. evaluated on only part of the data) and the second estimated based on the entire data set. This suggests that parts of your data are very different from other parts. What may help here is to run Auto-WEKA again with a different random seed so that its internal split into cross-validation folds will be different.

**Comment by federicoparra** [comment]
Hi. That doesn't seem to be the problem. If I run 10-fold cross-validation with Weka (if I choose the option in Classify to do 10 folds) with auto-weka classifier the folds (processed at the end after learning) have much more conservative results. There is no part of my dataset that could give those amazing results, I wish!!! :) These are only 240 cases with 500+ features, very complex to get any model to work - Google AutoMl gets only conservative results. It's definitely training and testing on the same data in terms of the reporting. Interestingly, the step validations reported below the window during auto-weka training have good conservative estimates.

**Comment by federicoparra** [comment]
To give you an idea what we are talking about here, most models including Google AutoML do not get better than .20 correlation with this dataset. So .7 or .9 can only happen if you are learning and testing on the same dataset.

**Comment by larskotthoff** [comment]
Hmm, it's hard to debug this without seeing the data. Have you tried running Auto-WEKA multiple times with different random seeds? It's possible that this one run was just extremely lucky.

**Comment by federicoparra** [comment]
Doesn't sound like it since different runs have suggested different models if I change the metric (to error for instance), and in every case they are too good to be true (and do not perform well on test data). I can send you the datafile if you agree to test it on your end, would that work? Can I have your email for that?

**Comment by larskotthoff** [comment]
I can certainly have a look, but no promises. larsko@uwyo.edu
